src/ctap2/commands/mod.rs

```diff
@@ -103,136 +102,137 @@ pub(crate) enum PinUvAuthResult {
     /// If Err() is returned, either cancel the operation or ask the user for a PIN
     fn determine_pin_uv_auth<D: FidoDevice>(
         &mut self,
         dev: &mut D,
         skip_uv: bool,
+        permission: PinUvAuthTokenPermission,
     ) -> Result<PinUvAuthResult, AuthenticatorError> {
+        // In case we error out in the middle
+        self.set_pin_auth(None);
+
         // CTAP1/U2F-only devices do not support PinUvAuth, so we skip it
         if !dev.supports_ctap2() {
             self.set_pin_auth(None);
             return Ok(PinUvAuthResult::DeviceIsCtap1);
         }
 
-        self.determine_pin_uv_auth_ctap_2_0(dev, skip_uv)
-    }
-
-    /// CTAP 2.0-only version:
-    /// "Getting pinUvAuthToken using getPinToken (superseded)"
-    fn determine_pin_uv_auth_ctap_2_0<D: FidoDevice>(
-        &mut self,
-        dev: &mut D,
-        skip_uv: bool,
-    ) -> Result<PinUvAuthResult, AuthenticatorError> {
         let info = dev
             .get_authenticator_info()
-            .ok_or(AuthenticatorError::Platform)?;
-
-        if self.pin().is_none() {
-            // Only use UV, if the device supports it and we don't skip it
-            // which happens as a fallback, if UV-usage failed too many times
-            let supports_uv = !skip_uv && info.options.user_verification == Some(true);
-            let supports_pin = info.options.client_pin.is_some();
-            let pin_configured = info.options.client_pin == Some(true);
+            .ok_or(AuthenticatorError::HIDError(HIDError::DeviceNotInitialized))?;
+
+        // Only use UV, if the device supports it and we don't skip it
+        // which happens as a fallback, if UV-usage failed too many times
+        // Note: In theory, we could also repeatedly query GetInfo here and check
+        //       if uv is set to Some(true), as tokens should set it to Some(false)
+        //       if UV is blocked (too many failed attempts). But the CTAP2.0-spec is
+        //       vague and I don't trust all tokens to implement it that way. So we
+        //       keep track of it ourselves, using `skip_uv`.
+        let supports_uv = !skip_uv && info.options.user_verification == Some(true);
+        let supports_pin = info.options.client_pin.is_some();
+        let pin_configured = info.options.client_pin == Some(true);
+
+        if self.get_uv_option() == Some(false) {
+            // TODO: API needs a better way to express "uv = discouraged"
+            let uv = if Self::command() == Command::MakeCredentials {
+                Some(false)
+            } else {
+                None
+            };
+            self.set_uv_option(uv);
+            return Ok(PinUvAuthResult::NoAuthRequired);
+        }
 
-            // If the device supports internal user-verification (e.g. fingerprints),
-            // skip PIN-stuff
+        // CTAP 2.1
+        let (res, pin_auth_token) = if info.options.pin_uv_auth_token == Some(true) {
```

Comment:
A simpler way to structure this is to handle all of the error cases first and then determine which getPinUvAuthToken variant to use. You can inline the functions from `transport/mod.rs` here so that they share the `establish_shared_secret` step (and unify that with the `UsingInternalUv` case). And then you can `let pin_auth_token = dev.send_cbor(...)?` and call `repackage_pin_errors` from a single location.

Reply:
With reference to my comment from above about moving functions from `FIDODevice`: I think inlining the functions would be unhelpful in the long-run, as we need them elsewhere.
I tried to condense the structure and pull some errors out, while removing superfluous ones (e.g. `PinRequired` is already handled by `get_pin_token()` and doesn't need to be repeated here).
And I also simply return the result and call `repackage_pin_errors()` on that result in one place afterwards, instead of everywhere.
Not what you had in mind, but maybe a workable middle ground :)